Mathlib/NumberTheory/LSeries/AbstractFuncEq.lean

```diff
@@ -0,0 +1,471 @@
+there is no `good` choice of value at `1`. -/
+def f_modif : ℝ → E :=
+  (Ioi 1).indicator (fun x ↦ P.f x - P.f₀) +
+  (Ioo 0 1).indicator (fun x ↦ P.f x - (P.ε * ↑(x ^ (-P.k))) • P.g₀)
+
+/-- Piecewise modified version of `g` with optimal asymptotics. -/
+def g_modif : ℝ → E :=
+  (Ioi 1).indicator (fun x ↦ P.g x - P.g₀) +
+  (Ioo 0 1).indicator (fun x ↦ P.g x - (P.ε⁻¹ * ↑(x ^ (-P.k))) • P.f₀)
+
+lemma hf_modif_int :
+    LocallyIntegrableOn P.f_modif (Ioi 0) := by
+  have : LocallyIntegrableOn (fun x : ℝ ↦ (P.ε * ↑(x ^ (-P.k))) • P.g₀) (Ioi 0) := by
+    refine ContinuousOn.locallyIntegrableOn ?_ measurableSet_Ioi
+    refine ContinuousAt.continuousOn (fun x (hx : 0 < x) ↦ ?_)
+    refine (continuousAt_const.mul ?_).smul continuousAt_const
+    exact continuous_ofReal.continuousAt.comp (continuousAt_rpow_const _ _ (Or.inl hx.ne'))
+  refine LocallyIntegrableOn.add (fun x hx ↦ ?_) (fun x hx ↦ ?_)
+  · obtain ⟨s, hs, hs'⟩ := P.hf_int.sub (locallyIntegrableOn_const _) x hx
+    refine ⟨s, hs, ?_⟩
+    rw [IntegrableOn, integrable_indicator_iff measurableSet_Ioi, IntegrableOn,
+      Measure.restrict_restrict measurableSet_Ioi, ← IntegrableOn]
+    exact hs'.mono_set (Set.inter_subset_right _ _)
+  · obtain ⟨s, hs, hs'⟩ := P.hf_int.sub this x hx
+    refine ⟨s, hs, ?_⟩
+    rw [IntegrableOn, integrable_indicator_iff measurableSet_Ioo, IntegrableOn,
+      Measure.restrict_restrict measurableSet_Ioo, ← IntegrableOn]
+    exact hs'.mono_set (Set.inter_subset_right _ _)
+
+lemma hf_modif_FE (x : ℝ) (hx : 0 < x) :
+    P.f_modif (1 / x) = (P.ε * ↑(x ^ P.k)) • P.g_modif x := by
+  rcases lt_trichotomy 1 x with hx' | rfl | hx'
+  · have : 1 / x < 1 := by rwa [one_div_lt hx one_pos, div_one]
+    rw [f_modif, Pi.add_apply, indicator_of_not_mem (not_mem_Ioi.mpr this.le),
+      zero_add, indicator_of_mem (mem_Ioo.mpr ⟨div_pos one_pos hx, this⟩), g_modif, Pi.add_apply,
+      indicator_of_mem (mem_Ioi.mpr hx'), indicator_of_not_mem
+      (not_mem_Ioo_of_ge hx'.le), add_zero, P.h_feq _ hx, smul_sub]
+    simp_rw [rpow_neg (one_div_pos.mpr hx).le, one_div, inv_rpow hx.le, inv_inv]
+  · rw [div_one, f_modif, Pi.add_apply, indicator_of_not_mem, indicator_of_not_mem,
+      g_modif, Pi.add_apply, indicator_of_not_mem, indicator_of_not_mem]
+    simp only [add_zero, smul_zero]
+    all_goals { simp }
+  · have : 1 < 1 / x := by rwa [lt_one_div one_pos hx, div_one]
+    rw [f_modif, Pi.add_apply, indicator_of_mem (mem_Ioi.mpr this),
+      indicator_of_not_mem (not_mem_Ioo_of_ge this.le), add_zero, g_modif, Pi.add_apply,
+      indicator_of_not_mem (not_mem_Ioi.mpr hx'.le),
+      indicator_of_mem (mem_Ioo.mpr ⟨hx, hx'⟩), zero_add, P.h_feq _ hx, smul_sub]
+    simp_rw [rpow_neg hx.le, ← mul_smul]
+    field_simp [(rpow_pos_of_pos hx P.k).ne', P.hε]
+
+/-- Given a weak FE-pair `(f, g)`, modify it into a strong FE-pair by subtracting suitable
+correction terms from `f` and `g`. -/
+def modifStrongFEPair : StrongFEPair E where
```

Comment:
Why not call it simply `strongFEPair`? The name looks like the strong FE-pair is somehow modified, but what the function does is to associate a strong pair to a weak one (and I assume there is only one reasonable way of doing that).

Reply:
I wanted to make it clear in the name that the resulting strong FE pair does *not* have the same underlying functions as the weak FE pair does – it's different underlying data, not the same data repackaged. There is definitely a lot of choice in how one constructs a strong FE pair from a weak one: strong FE-pairs are a codimension 2 subspace in the space of weak FE-pairs, and we are choosing a specific complementary subspace.

Reply:
OK; I can see that one could, e.g., choose more general splitting points `t` and `1/t` instead of `1` and `1` in the definition of the modified functions. Still, I don't quite like the name and would prefer that `modif` is removed. I think the "making clear that" is better done in the docstring.

Reply:
Do you have an alternative suggestion for a naming scheme? If `WeakFEPair.modifStrongFEPair` becoms `WeakFEPair.strongFEPair`, then what should the lemmas needed in building this object, e.g. `WeakFEPair.hf_modif_int`, be called?

Reply:
I think it's OK to leave the other names as they are, as they indeed talk about modified versions of the data. I would consider them implementation details of the projection to strong pairs, which is the main goal here. The difference is what `modif` appears to refer to.

Maybe `WeakFEPair.toStrongFEPair` captures a bit better what this is intended to do?

Reply:
I have changed it to `WeakFEPair.toStrongFEPair`, while leaving the preliminary lemmas unchanged, as suggested.